// @kontsedal/locco/dist/index.js
// src/errors.ts
var LoccoError = class extends Error {
  name;
  constructor(description, name = "LoccoError") {
    super(description);
    Object.setPrototypeOf(this, new.target.prototype);
    this.name = name;
    if (typeof Error.captureStackTrace === "function") {
      Error.captureStackTrace(this);
    }
  }
};
var LockCreateError = class extends LoccoError {
  constructor(description = "") {
    super(description, "LockCreateError");
  }
};
var LockReleaseError = class extends LoccoError {
  constructor(description = "") {
    super(description, "LockReleaseError");
  }
};
var LockExtendError = class extends LoccoError {
  constructor(description = "") {
    super(description, "LockExtendError");
  }
};
var RetryError = class extends LoccoError {
  constructor(description = "") {
    super(description, "RetryError");
  }
};
var ValidationError = class extends LoccoError {
  constructor(description = "") {
    super(description, "ValidationError");
  }
};

// src/utils/validators.ts
function isPositiveInteger(value) {
  return typeof value === "number" && Number.isInteger(value) && value > 0;
}
function isUndefinedOrPositiveInteger(value) {
  return isUndefined(value) || isPositiveInteger(value);
}
function isUndefinedOrFunction(value) {
  return isUndefined(value) || isFunction(value);
}
function isFunction(value) {
  return typeof value === "function";
}
function isObject(value) {
  return typeof value === "object" && value !== null;
}
function isUndefined(value) {
  return typeof value === "undefined";
}
function isMongoError(value) {
  return value instanceof Error && "code" in value;
}
function isStringWithContents(value) {
  return typeof value === "string" && value.length > 0;
}
function validateRetrySettings(settings) {
  if (!settings || typeof settings !== "object") {
    throw new ValidationError("Retry settings should be an object");
  }
  const { retryDelayFn, retryDelay, retryTimes, totalTime } = settings;
  if (!isUndefinedOrPositiveInteger(retryDelay)) {
    throw new ValidationError("retryDelay should be an integer");
  }
  if (!isUndefinedOrPositiveInteger(retryTimes)) {
    throw new ValidationError("retryTimes should be an integer");
  }
  if (!isUndefinedOrPositiveInteger(totalTime)) {
    throw new ValidationError("totalTime should be an integer");
  }
  if (!isUndefinedOrFunction(retryDelayFn)) {
    throw new ValidationError("retryDelayFn should be a function");
  }
  if (isFunction(retryDelayFn) && isPositiveInteger(retryDelay)) {
    throw new ValidationError("Can't have both retryDelayFn and retryDelay");
  }
  if (isUndefined(retryDelayFn) && (isUndefined(retryTimes) || isUndefined(retryDelay))) {
    throw new ValidationError(
      "retryTimes and retryDelay should be specified if retryDelayFn is not provided"
    );
  }
}
function validateTtl(ttl) {
  if (!isPositiveInteger(ttl)) {
    throw new ValidationError("Ttl should be a positive integer");
  }
}
function validateKey(key) {
  if (!isStringWithContents(key)) {
    throw new ValidationError(
      "Key should be a string with at least one character"
    );
  }
}
function validateUniqueValue(uniqueValue) {
  if (!isStringWithContents(uniqueValue)) {
    throw new ValidationError(
      "uniqueValue should be a string with at least one character"
    );
  }
}
function validateAdapter(adapter) {
  if (!isObject(adapter)) {
    throw new ValidationError("Adapter is required");
  }
  if (!isFunction(adapter.createLock) || !isFunction(adapter.releaseLock) || !isFunction(adapter.extendLock) || !isFunction(adapter.isValidLock)) {
    throw new ValidationError("Adapter is invalid");
  }
}

// src/utils/wait.ts
var wait = (ms) => {
  return new Promise((res) => setTimeout(res, ms));
};

// src/utils/retry.ts
var retry = ({
  settings,
  fn,
  shouldProceedFn
}) => {
  validateRetrySettings(settings);
  let attemptNumber = -1;
  const startedAt = Date.now();
  const tick = async (previousDelay = 0) => {
    if (settings.totalTime !== void 0 && Date.now() - startedAt >= settings.totalTime) {
      throw new RetryError("Total time exceeded");
    }
    attemptNumber += 1;
    try {
      return await fn();
    } catch (error) {
      const shouldProceed = shouldProceedFn(error);
      if (!shouldProceed) {
        throw error;
      }
      if (!isUndefined(settings.retryTimes) && settings.retryTimes <= attemptNumber + 1) {
        throw new RetryError("Reached retry times limit");
      }
      let delay;
      if (settings.retryDelayFn) {
        delay = await settings.retryDelayFn({
          attemptNumber,
          startedAt,
          settings,
          previousDelay,
          stop: () => {
            throw new RetryError("Was manually stopped");
          }
        });
      } else {
        delay = settings.retryDelay;
      }
      await wait(delay);
      return tick(delay);
    }
  };
  return tick();
};

// src/utils/getRandomHash.ts
import { randomBytes } from "crypto";
var getRandomHash = () => {
  return randomBytes(16).toString("hex");
};

// src/lock.ts
var Lock = class _Lock {
  adapter;
  key;
  retrySettings;
  uniqueValue;
  ttl;
  locked = false;
  acquiring = false;
  released = false;
  constructor({
    adapter,
    key,
    retrySettings,
    uniqueValue,
    ttl
  }) {
    validateRetrySettings(retrySettings);
    validateTtl(ttl);
    validateKey(key);
    this.retrySettings = retrySettings;
    this.adapter = adapter;
    this.key = key;
    this.ttl = ttl;
    this.uniqueValue = uniqueValue ?? getRandomHash();
  }
  async acquire(cb) {
    if (this.locked || this.acquiring) {
      throw new LoccoError("Lock is already acquired");
    }
    this.acquiring = true;
    try {
      await retry({
        settings: this.retrySettings,
        fn: () => this.adapter.createLock({
          key: this.key,
          uniqueValue: this.uniqueValue,
          ttl: this.ttl
        }),
        shouldProceedFn: (error) => error instanceof LockCreateError
      });
      this.locked = true;
    } catch (error) {
      this.acquiring = false;
      throw error;
    }
    if (isFunction(cb)) {
      try {
        return await cb(this);
      } finally {
        await this.release();
      }
    }
    return this;
  }
  async release({ throwOnFail = false } = {}) {
    try {
      if (this.released) {
        throw new LoccoError(
          "Locco:::Lock:::release Can't release resource twice"
        );
      }
      await this.adapter.releaseLock({
        key: this.key,
        uniqueValue: this.uniqueValue
      });
      this.released = true;
    } catch (error) {
      if (throwOnFail) {
        throw error;
      }
    }
  }
  async isLocked() {
    return this.adapter.isValidLock({
      key: this.key,
      uniqueValue: this.uniqueValue
    });
  }
  async extend(ttl) {
    if (!this.locked) {
      throw new LoccoError(
        "Locco:::Lock:::extend Can't extend a lock before a lock success"
      );
    }
    if (this.released) {
      throw new LoccoError(
        "Locco:::Lock:::extend Can't extend a released lock"
      );
    }
    validateTtl(ttl);
    return this.adapter.extendLock({
      key: this.key,
      uniqueValue: this.uniqueValue,
      ttl
    });
  }
  setRetrySettings(settings) {
    if (this.locked) {
      throw new LoccoError(
        "Locco:::setRetrySettings Can't change retry settings after lock success"
      );
    }
    validateRetrySettings(settings);
    return new _Lock({
      adapter: this.adapter,
      key: this.key,
      ttl: this.ttl,
      retrySettings: settings,
      uniqueValue: this.uniqueValue
    });
  }
};

// src/locker.ts
var Locker = class {
  adapter;
  retrySettings;
  constructor({
    adapter,
    retrySettings
  }) {
    validateAdapter(adapter);
    validateRetrySettings(retrySettings);
    this.adapter = adapter;
    this.retrySettings = retrySettings;
  }
  lock(key, ttl) {
    validateTtl(ttl);
    validateKey(key);
    return new Lock({
      adapter: this.adapter,
      key,
      ttl,
      retrySettings: this.retrySettings
    });
  }
};

// src/adapters/inMemoryAdapter.ts
var InMemoryAdapter = class {
  storage = /* @__PURE__ */ new Map();
  timers = /* @__PURE__ */ new Map();
  async createLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const entry = this.storage.get(key);
    if (entry && entry.expireAt > Date.now()) {
      throw new LockCreateError();
    }
    this.setLock({ key, uniqueValue, ttl });
  }
  async releaseLock({
    key,
    uniqueValue
  }) {
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const entry = this.storage.get(key);
    const now = Date.now();
    if (!entry || entry.expireAt <= now) {
      throw new LockReleaseError("Lock is already expired");
    }
    if (entry.uniqueValue !== uniqueValue) {
      throw new LockReleaseError("Lock is already taken");
    }
    this.storage.delete(key);
    const timer = this.timers.get(key);
    if (timer) {
      clearTimeout(timer);
      this.timers.delete(key);
    }
  }
  async extendLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const entry = this.storage.get(key);
    if (!entry || entry.expireAt <= Date.now() || entry.uniqueValue !== uniqueValue) {
      throw new LockExtendError();
    }
    this.setLock({ key, uniqueValue, ttl });
  }
  async isValidLock({
    key,
    uniqueValue
  }) {
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const entry = this.storage.get(key);
    return Boolean(
      entry && entry.uniqueValue === uniqueValue && entry.expireAt > Date.now()
    );
  }
  setLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const expireAt = Date.now() + ttl;
    this.storage.set(key, { uniqueValue, expireAt });
    const existingTimer = this.timers.get(key);
    if (existingTimer) {
      clearTimeout(existingTimer);
    }
    const timer = setTimeout(() => {
      const entryAfterTime = this.storage.get(key);
      if (entryAfterTime && entryAfterTime.uniqueValue === uniqueValue && entryAfterTime.expireAt === expireAt) {
        this.storage.delete(key);
      }
      this.timers.delete(key);
    }, ttl);
    timer.unref();
    this.timers.set(key, timer);
  }
};

// src/adapters/ioRedisAdapter.ts
var RELEASE_LOCK_SCRIPT = `
  if redis.call("get",KEYS[1]) == ARGV[1] then
    return redis.call("del",KEYS[1])
  else
    return 0
  end
`;
var EXTEND_LOCK_SCRIPT = `
  if redis.call("get", KEYS[1]) == ARGV[1] then
    return redis.call("set", KEYS[1], ARGV[1], "PX", ARGV[2])
  else
    return nil
  end
`;
var IoRedisAdapter = class {
  client;
  constructor({ client }) {
    client.defineCommand("releaseLock", {
      numberOfKeys: 1,
      lua: RELEASE_LOCK_SCRIPT
    });
    client.defineCommand("extendLock", {
      numberOfKeys: 1,
      lua: EXTEND_LOCK_SCRIPT
    });
    this.client = client;
  }
  async createLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const result = await this.client.set(key, uniqueValue, "PX", ttl, "NX");
    if (result !== "OK") {
      throw new LockCreateError();
    }
  }
  async releaseLock({
    key,
    uniqueValue
  }) {
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const result = await this.client.releaseLock(key, uniqueValue);
    if (result === 1) {
      return;
    }
    throw new LockReleaseError();
  }
  async extendLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const result = await this.client.extendLock(key, uniqueValue, ttl);
    if (result === "OK") {
      return;
    }
    throw new LockExtendError();
  }
  async isValidLock({
    key,
    uniqueValue
  }) {
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const entry = await this.client.get(key);
    return Boolean(entry && entry === uniqueValue);
  }
};

// src/adapters/mongoAdapter.ts
var MONGO_DUPLICATE_ERROR_CODE = 11e3;
var MongoAdapter = class {
  collection;
  indexesCreated = false;
  indexesPromise = null;
  constructor({
    client,
    locksCollectionName = "locco-locks",
    dbName
  }) {
    this.collection = client.db(dbName).collection(locksCollectionName);
  }
  async createIndexes() {
    if (this.indexesCreated) {
      return;
    }
    if (!this.indexesPromise) {
      this.indexesPromise = Promise.all([
        this.collection.createIndex({ key: 1 }, { unique: true }),
        this.collection.createIndex({ expireAt: 1 }, { expireAfterSeconds: 0 }),
        this.collection.createIndex(
          { key: 1, expireAt: 1, uniqueValue: 1 },
          {}
        )
      ]).then(() => {
        this.indexesCreated = true;
      });
    }
    return this.indexesPromise;
  }
  async createLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    await this.createIndexes();
    try {
      await this.collection.updateOne(
        {
          key,
          expireAt: { $lt: /* @__PURE__ */ new Date() }
        },
        { $set: { key, uniqueValue, expireAt: new Date(Date.now() + ttl) } },
        { upsert: true }
      );
    } catch (error) {
      if (isMongoError(error) && (error == null ? void 0 : error.code) === MONGO_DUPLICATE_ERROR_CODE) {
        throw new LockCreateError();
      }
      throw error;
    }
  }
  async releaseLock({
    key,
    uniqueValue
  }) {
    validateKey(key);
    validateUniqueValue(uniqueValue);
    await this.createIndexes();
    const result = await this.collection.deleteOne({
      key,
      expireAt: { $gt: /* @__PURE__ */ new Date() },
      uniqueValue
    });
    if (result.deletedCount === 0) {
      throw new LockReleaseError();
    }
  }
  async extendLock({
    key,
    uniqueValue,
    ttl
  }) {
    validateTtl(ttl);
    validateKey(key);
    validateUniqueValue(uniqueValue);
    await this.createIndexes();
    const result = await this.collection.updateOne(
      {
        key,
        expireAt: { $gt: /* @__PURE__ */ new Date() },
        uniqueValue
      },
      { $set: { expireAt: new Date(Date.now() + ttl) } }
    );
    if (result.matchedCount === 0) {
      throw new LockExtendError();
    }
  }
  async isValidLock({
    key,
    uniqueValue
  }) {
    validateKey(key);
    validateUniqueValue(uniqueValue);
    const entry = await this.collection.findOne({
      key,
      expireAt: { $gt: /* @__PURE__ */ new Date() },
      uniqueValue
    });
    return Boolean(entry);
  }
};
export {
  InMemoryAdapter,
  IoRedisAdapter,
  LoccoError,
  Lock,
  LockCreateError,
  LockExtendError,
  LockReleaseError,
  Locker,
  MongoAdapter,
  RetryError,
  ValidationError
};
